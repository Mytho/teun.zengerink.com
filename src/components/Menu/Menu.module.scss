@import '../../styles/index';

.menu {
  h1 {
    position: fixed;
    z-index: 4;
    top: $margin;
    left: $margin;
    padding-top: 4px;
  }
}

.navigation {
  @include transition(0.1s);

  position: fixed;
  top: 0;
  left: $margin-small;
  bottom: 0;
  transform: translateY(-50px);
  visibility: hidden;
  opacity: 0;
  z-index: 3;
  padding: 60px $margin $margin $margin;
  width: 100%;
  background: $background-color-transparent;

  h2 {
    margin-top: 30px;
    margin-bottom: 10px;
    cursor: default;
  }

  ul {
    list-style: none;
    padding: 0;
  }

  li a {
    font-size: 0.8em;
  }
}

.hamburger {
  display: block;
  z-index: 4;
  position: fixed;
  top: $margin;
  right: $margin;
  cursor: pointer;
}

.line {
  @include transition(0.3s);

  display: block;
  width: 24px;
  height: 3px;
  margin: 4px auto;
  background-color: $text-color;
}

.socials {
  position: absolute;
  right: $margin;
  padding-right: $margin;
  margin-right: $margin;
  bottom: $margin;

  a {
    text-decoration: none;
  }

  img {
    float: left;
    width: 22px;
    height: 22px;
    opacity: 0.7;

    &:hover {
      opacity: 1;
    }
  }
}

.open {
  .navigation {
    visibility: visible;
    opacity: 1;
    transform: none;
  }

  .hamburger .line:nth-child(1) {
    -webkit-transform: translateY(8px) rotate(45deg);
    -ms-transform: translateY(8px) rotate(45deg);
    -o-transform: translateY(8px) rotate(45deg);
    transform: translateY(8px) rotate(45deg);
  }

  .hamburger .line:nth-child(2) {
    opacity: 0;
  }

  .hamburger .line:nth-child(3) {
    -webkit-transform: translateY(-6px) rotate(-45deg);
    -ms-transform: translateY(-6px) rotate(-45deg);
    -o-transform: translateY(-6px) rotate(-45deg);
    transform: translateY(-6px) rotate(-45deg);
  }
}

@include from-tablet-portrait {
  .menu {
    h1 {
      padding-top: 0;
    }
  }

  .navigation {
    position: fixed;
    opacity: 1;
    visibility: visible;
    right: $margin;
    bottom: auto;
    left: auto;
    transform: none;
    width: auto;
    padding: 0;
    text-align: right;

    h2 {
      margin-top: $margin;
      margin-bottom: 0;
      text-transform: uppercase;
    }

    ul {
      display: none;
      margin-top: 0;
      padding-top: $margin;
      padding-bottom: $margin;

      &:hover {
        display: block;
      }
    }

    h2:hover + ul {
      display: block;
    }
  }

  .hamburger {
    display: none;
  }

  .socials {
    position: fixed;
    left: auto;
    bottom: $margin;
    right: $margin;
    padding-right: 0;
    margin-right: 0;
  }
}
