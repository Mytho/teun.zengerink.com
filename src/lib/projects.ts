import fs from 'fs'
import path from 'path'
import slugify from 'slugify'
import sizeOf from 'image-size'
import yaml from 'js-yaml'

interface ConfigCaption {
  key: string
  caption: string
}

interface ConfigItem {
  title: string
  statement?: string
  captions?: ConfigCaption[]
}

export interface Photo {
  key: string
  url: string
  size: { width: number; height: number }
  caption?: string
}

export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
}

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), 'public/photos', slug)
  const fileNames = fs.readdirSync(photosDirectory)

  return fileNames.map((filename) => {
    const key = path.basename(filename, '.jpg')
    const caption = captions?.find((c) => c.key == key)

    return {
      key,
      caption: caption?.caption ?? '',
      url: `/photos/${slug}/${filename}`,
      size: sizeOf(path.join(photosDirectory, filename)),
    }
  })
}

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true })

    return {
      slug,
      title: project.title,
      statement: project.statement ?? '',
      photos: generatePhotos(slug, project.captions),
    }
  })
}

export const getProjects = async (): Promise<Project[]> => {
  try {
    const file = path.join(process.cwd(), 'src', 'config.yml')
    const config = yaml.load(fs.readFileSync(file, 'utf-8'))
    return await generateProjects(config)
  } catch (e) {
    console.error(e)
  }
}
