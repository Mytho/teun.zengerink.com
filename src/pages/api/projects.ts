import fs from "fs";
import { IncomingMessage, ServerResponse } from 'http'
import path from "path";
import slugify from "slugify";

interface ConfigCaption {
  key: string;
  caption: string;
}

interface ConfigItem {
  title: string;
  statement?: string;
  captions?: ConfigCaption[];
}

interface Photo {
  key: string;
  url: string;
  caption?: string;
}

export interface Project {
  title: string;
  slug: string;
  photos: Photo[];
  statement?: string;
}

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), "public/photos", slug);
  const fileNames = fs.readdirSync(photosDirectory);

  return fileNames.map((filename) => {
    const key = path.basename(filename, ".jpg");
    const caption = captions?.find((c) => c.key == key)?.caption;

    return {
      key,
      caption,
      url: `photos/${slug}/${filename}`,
    };
  });
};

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true });

    return {
      title: project.title,
      statement: project.statement,
      slug,
      photos: generatePhotos(slug, project.captions),
    };
  });
};

const config: ConfigItem[] = [
  {
    title: "Take, Make, Waste",
    statement: "TODO",
    captions: [
      {
        key: "01",
        caption: "Image 01: TODO",
      },
      {
        key: "02",
        caption: "Image 02: TODO",
      },
      {
        key: "03",
        caption: "Image 03: TODO",
      },
    ],
  },
  {
    title: "Point of no Return",
    statement: "TODO",
  },
  {
    title: "Under Your Nose",
    statement: "TODO",
  },
  {
    title: "Fortress",
    statement: "TODO",
  },
  {
    title: "Symbiosis",
    statement: "TODO",
  },
  {
    title: "Untitled",
  },
];

const handler = (request: IncomingMessage, response: ServerResponse): void => {
  response.statusCode = 200;
  response.setHeader("Content-Type", "application/json");
  response.end(JSON.stringify({ projects: generateProjects(config) }));
};

export default handler;
