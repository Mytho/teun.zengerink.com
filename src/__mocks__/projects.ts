import { Project } from '../lib/projects'

const projects: Project[] = [
  {
    title: 'My First Project',
    slug: 'my-first-project',
    statement: 'The statement for the project',
    photos: [
      {
        key: '01',
        url: '/photos/my-first-project/01.jpg',
        size: { width: 800, height: 600 },
      },
      {
        key: '02',
        url: '/photos/my-first-project/02.jpg',
        size: { width: 600, height: 800 },
      },
      {
        key: '03',
        url: '/photos/my-first-project/03.jpg',
        size: { width: 800, height: 600 },
      },
    ],
  },
  {
    title: 'Project without a statement',
    slug: 'no-statement',
    photos: [
      {
        key: '01',
        url: '/photos/no-statement/01.jpg',
        caption: 'This is image 1',
        size: { width: 800, height: 600 },
      },
      {
        key: '02',
        url: '/photos/no-statement/02.jpg',
        caption: 'This is image 2',
        size: { width: 800, height: 600 },
      },
      {
        key: '03',
        url: '/photos/no-statement/03.jpg',
        caption: 'This is image 3',
        size: { width: 600, height: 800 },
      },
      {
        key: '04',
        url: '/photos/no-statement/04.jpg',
        caption: 'This is image 4',
        size: { width: 800, height: 600 },
      },
    ],
  },
]

export default projects
