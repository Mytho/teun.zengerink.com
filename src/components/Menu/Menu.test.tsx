import { fireEvent, render } from '@testing-library/react'
import React from 'react'
import projects from '../../__mocks__/projects'
import Menu from './Menu'

const defaultProps = { pageTitle: 'My Homepage', projects }

describe('Menu', () => {
  const renderComponent = (props = defaultProps) => render(<Menu {...props} />)

  it('renders correctly', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('toggles the menu when the hamburger is clicked', () => {
    const { getByTestId } = renderComponent()
    const menu = getByTestId('menu')
    const hamburger = getByTestId('hamburger')
    expect(menu.classList.contains('open')).toBe(false)
    fireEvent.click(hamburger)
    expect(menu.classList.contains('open')).toBe(true)
    fireEvent.click(hamburger)
    expect(menu.classList.contains('open')).toBe(false)
  })
})
