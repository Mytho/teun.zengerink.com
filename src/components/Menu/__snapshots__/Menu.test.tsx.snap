// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly when on a project page 1`] = `
<div>
  <aside
    class="menu"
    data-testid="menu"
  >
    <a
      href="/"
    >
      <h1>
        My Homepage
      </h1>
    </a>
    <div
      class="hamburger"
      data-testid="hamburger"
    >
      <span
        class="line"
      />
      <span
        class="line"
      />
      <span
        class="line"
      />
    </div>
    <nav
      class="navigation"
    >
      <h2>
        Projects
      </h2>
      <ul>
        <li
          class=""
        >
          <a
            href="/work/my-first-project/01"
          >
            My First Project
          </a>
        </li>
        <li
          class="active"
        >
          <a
            href="/work/no-statement/01"
          >
            Project without a statement
          </a>
        </li>
      </ul>
      <span
        class="socials"
      >
        <a
          href="https://instagram.com/t.zengerink/"
        >
          <img
            alt="Instagram"
            src="/img/instagram.jpg"
          />
        </a>
      </span>
    </nav>
  </aside>
</div>
`;

exports[`Menu renders correctly when on homepage 1`] = `
<div>
  <aside
    class="menu"
    data-testid="menu"
  >
    <a
      href="/"
    >
      <h1>
        My Homepage
      </h1>
    </a>
    <div
      class="hamburger"
      data-testid="hamburger"
    >
      <span
        class="line"
      />
      <span
        class="line"
      />
      <span
        class="line"
      />
    </div>
    <nav
      class="navigation"
    >
      <h2>
        Projects
      </h2>
      <ul>
        <li
          class=""
        >
          <a
            href="/work/my-first-project/01"
          >
            My First Project
          </a>
        </li>
        <li
          class=""
        >
          <a
            href="/work/no-statement/01"
          >
            Project without a statement
          </a>
        </li>
      </ul>
      <span
        class="socials"
      >
        <a
          href="https://instagram.com/t.zengerink/"
        >
          <img
            alt="Instagram"
            src="/img/instagram.jpg"
          />
        </a>
      </span>
    </nav>
  </aside>
</div>
`;
